### PR TITLE
feat: pino logging

### PR DESCRIPTION
**Pull request by trentm**
This switches the APM agent to use pino for structured logging. Visible
changes:

- Default logging output is now JSON in ecs-logging format to stdout,
  rather than just a plain message to stdout or stderr (warn and above).
- Custom loggers are still supported and there shouldn't be any breakage
  there. This is done by wrapping the custom logger in a pino logger
  that will just pass the log message through. This means that when we
  add structure log fields (for example "foo" in `log.info({foo:
  bar}, 'my message')`), then users of a custom logger will not see
  those structured fields. I think that is fine.
- Added support for `ELASTIC_APM_LOGGER=false` to ignore a given custom logger to be used for "Debug mode" troubleshooting logging... i.e. to avoid not getting structured log fields in debug output from users for the (hopefully rare) cases where they are using a custom logger.

Note that I want this improved logging as a part of the
https://github.com/elastic/apm-nodejs-http-client/issues/136 work.

TODO:
- [x] The logLevel config var now *applies* to a given custom logger.
  Discuss this. Custom loggers are wrapped in a pino logger that
  writes the log message to the custom logger. The logLevel is on that
  pino logger wrapper. I think I should change this to set the wrapper
  to 'trace' level and *not* use `logLevel` as before... otherwise the
  user has *two* confusing log levels to coordinate.
- [x] Drop the 'pid' and 'hostname' default fields as discussed in #1430
  that this PR is obsoleting. I need to fix a bug in
  @elastic/ecs-format-pino for this.
- [x] Check/handle this case:
  `this.logger.warn(new Error('cannot flush agent before it is started'))`
- [x] Do we need `crlf: platform() === 'win32',` from the prev PR?
  I am inclined to *no* unless there is specific argument that it is required. My memory from Windows land was that basically everything except Notepad was fine with LF line endings. If that is the only reason, then no point in special casing here.
- [x] Update troubleshooting doc for logging changes. Including turning
  *off* a custom logger than can hide data.

Things to do after this PR:
- pass the logger to the http client
- change over lots of the current logging statements to use structured
  data, including a 'span' serializer (as in PR #1430)

### Checklist

- [x] Implement code
- [x] Add tests
- [x] Update TypeScript typings (N/A)
- [x] Update documentation
- [x] Add CHANGELOG.asciidoc entry
- [x] Commit message follows [commit guidelines](https://github.com/elastic/apm-agent-nodejs/blob/master/CONTRIBUTING.md#commit-message-guidelines)

### CODE REVIEW

**Comment by trentm** [comment]
https://gist.github.com/trentm/4bafafb24aabc37f55b571858c972f27 demos the various logging configs.


# SCENARIO=0 logging capabilities before this change

```
% SCENARIO=0 node demo-apm-agent-pino-logging.js
could not extract route name from request { foo: 'bar', answer: 42 }
hi Bob
```

# SCENARIO=1 default ecs-logging format

```
% SCENARIO=1 node demo-apm-agent-pino-logging.js
{"log.level":"info","@timestamp":"2021-03-16T00:38:01.485Z","ecs":{"version":"1.6.0"},"process":{"pid":22850},"host":{"hostname":"pink.local"},"log":{"logger":"elastic-apm-node"},"message":"could not extract route name from request {\"foo\":\"bar\",\"answer\":42}"}
{"log.level":"warn","@timestamp":"2021-03-16T00:38:01.486Z","ecs":{"version":"1.6.0"},"process":{"pid":22850},"host":{"hostname":"pink.local"},"log":{"logger":"elastic-apm-node"},"message":"hi"}

{"log.level":"warn","@timestamp":"2021-03-16T00:38:01.487Z","ecs":{"version":"1.6.0"},"process":{"pid":22850},"host":{"hostname":"pink.local"},"log":{"logger":"elastic-apm-node"},"error":{"type":"Error","message":"boom","stack_trace":"Error: boom\n    at Object.<anonymous> (/Users/trentm/tmp/p2/demo-apm-agent-pino-logging.js:106:20)\n    at Module._compile (internal/modules/cjs/loader.js:778:30)\n    at Object.Module._extensions..js (internal/modules/cjs/loader.js:789:10)\n    at Module.load (internal/modules/cjs/loader.js:653:32)\n    at tryModuleLoad (internal/modules/cjs/loader.js:593:12)\n    at Function.Module._load (internal/modules/cjs/loader.js:585:3)\n    at Function.Module.runMain (internal/modules/cjs/loader.js:831:12)\n    at startup (internal/bootstrap/node.js:283:19)\n    at bootstrapNodeJSCore (internal/bootstrap/node.js:623:3)"},"message":"oops"}
{"log.level":"info","@timestamp":"2021-03-16T00:38:01.487Z","ecs":{"version":"1.6.0"},"process":{"pid":22850},"host":{"hostname":"pink.local"},"log":{"logger":"elastic-apm-node"},"span":{"id":1},"a":1,"message":"started a span"}
{"log.level":"info","@timestamp":"2021-03-16T00:38:01.488Z","ecs":{"version":"1.6.0"},"process":{"pid":22850},"host":{"hostname":"pink.local"},"log":{"logger":"elastic-apm-node"},"module":"this-is-the-client","foo":"bar","message":"hi from the http client"}
```

which can be pretty-printed (and filtered) with my fledgling [`ecslog` tool](https://github.com/trentm/go-ecslog):

Screen Shot 2021-03-15 at 5 38 38 PM


# SCENARIO=2 a custom pino logger

```
% SCENARIO=2 node demo-apm-agent-pino-logging.js
{"level":30,"time":1615855276596,"pid":22936,"hostname":"pink.local","name":"myLogger","msg":"could not extract route name from request {\"foo\":\"bar\",\"answer\":42}"}
{"level":40,"time":1615855276597,"pid":22936,"hostname":"pink.local","name":"myLogger","msg":"hi"}

{"level":40,"time":1615855276598,"pid":22936,"hostname":"pink.local","name":"myLogger","err":{"type":"Error","message":"boom","stack":"Error: boom\n    at Object.<anonymous> (/Users/trentm/tmp/p2/demo-apm-agent-pino-logging.js:106:20)\n    at Module._compile (internal/modules/cjs/loader.js:778:30)\n    at Object.Module._extensions..js (internal/modules/cjs/loader.js:789:10)\n    at Module.load (internal/modules/cjs/loader.js:653:32)\n    at tryModuleLoad (internal/modules/cjs/loader.js:593:12)\n    at Function.Module._load (internal/modules/cjs/loader.js:585:3)\n    at Function.Module.runMain (internal/modules/cjs/loader.js:831:12)\n    at startup (internal/bootstrap/node.js:283:19)\n    at bootstrapNodeJSCore (internal/bootstrap/node.js:623:3)"},"msg":"oops"}
{"level":30,"time":1615855276598,"pid":22936,"hostname":"pink.local","name":"myLogger","span":{"id":1},"a":1,"msg":"started a span"}
{"level":30,"time":1615855276598,"pid":22936,"hostname":"pink.local","name":"myLogger","module":"this-is-the-client","foo":"bar","msg":"hi from the http client"}
```

This provides full fidelity. I.e. structured fields (like `a` above), serializers (like `err` and `span`), and child loggers are supported.

This can be pretty-printed with [pino-pretty](https://github.com/pinojs/pino-pretty)

# SCENARIO=3 a custom logger using console-log-level

A quick way to get the old logging format back.

```
% SCENARIO=3 node demo-apm-agent-pino-logging.js
could not extract route name from request {"foo":"bar","answer":42}
hi

oops
started a span
hi from the http client
```

Basically it works as before, but we can use the pino logger API in the agent code without breakage. The user of the custom logger should loses out on structured fields that we add.

# SCENARIO=4 using a bunyan logger

```
% SCENARIO=4 node demo-apm-agent-pino-logging.js
{"name":"myBunyanLogger","hostname":"pink.local","pid":23045,"level":30,"msg":"could not extract route name from request {\"foo\":\"bar\",\"answer\":42}","time":"2021-03-16T00:44:29.624Z","v":0}
{"name":"myBunyanLogger","hostname":"pink.local","pid":23045,"level":40,"msg":"hi","time":"2021-03-16T00:44:29.625Z","v":0}

{"name":"myBunyanLogger","hostname":"pink.local","pid":23045,"level":40,"msg":"oops","time":"2021-03-16T00:44:29.626Z","v":0}
{"name":"myBunyanLogger","hostname":"pink.local","pid":23045,"level":30,"msg":"started a span","time":"2021-03-16T00:44:29.626Z","v":0}
{"name":"myBunyanLogger","hostname":"pink.local","pid":23045,"level":30,"msg":"hi from the http client","time":"2021-03-16T00:44:29.626Z","v":0}
```

I only really included this because our current docs mention using a Bunyan custom logger. It works, but loses fidelity for new structured logging. (The bunyan API is similar enough to pino -- pino was based on it -- that we *could* keep fidelity with a custom Bunyan logger. However I propose that we do *not* sign up for that long term compat game.)

**Comment by trentm** [comment]
# format change impact on user

The default log level is "info", so with the change to pino format logging,
all info, warn, and error log statements will will be different for the default
case.

For example:

```
# Before
APM Server transport error (ECONNREFUSED): connect ECONNREFUSED 127.0.0.1:8200

# After (newline added to attempt simulate typical wrapping in the console)
{"log.level":"error","@timestamp":"2021-03-19T00:21:17.571Z","log":{"logger":"elastic-apm-node"},
"ecs":{"version":"1.6.0"},"message":"APM Server transport error (ECONNREFUSED): connect ECONNREFUSED 127.0.0.1:8200"}
```

Or, more nicely if/when `ecslog` gets used:

Screen Shot 2021-03-18 at 5 24 26 PM



Here are all the current `logger.info` cases:

```
% rg logger.info
lib/agent.js
418:      agent.logger.info('Sending error to Elastic APM: %o', { id })

lib/config.js
320:                agent.logger.info(`Remote config success. Updated logger with new logLevel: ${value}`)
322:              agent.logger.info(`Remote config success. Updated ${key}: ${value}`)
```

and `logger.warn` cases:

```
% rg logger.warn
lib/agent.js
    -   const err = new Error('cannot flush agent before it is started')
465:    this.logger.warn({ err }, err.message)

lib/instrumentation/http-shared.js
198:        agent.logger.warn('unable to identify http.ClientRequest url %o', { id: id })

lib/instrumentation/generic-span.js
101:    this._agent.logger.warn('Illegal characters used in tag key: %s', key)

lib/instrumentation/modules/fastify.js
74:    agent.logger.warn('Elastic APM cannot automaticaly capture errors on this verison of Fastify. Upgrade to version 2.0.0 or later.')

lib/config.js
309:            agent.logger.warn(`Remote config failure. Unsupported config names: ${unknown.join(', ')}`)
427:      logger.warn('Invalid "transactionSampleRate" config value %s, falling back to default %s',
451:      logger.warn('Invalid "cloudProvider" config value %s, falling back to default %s',
614:      logger.warn('unrecognized boolean value "%s" for "%s"', value, key)

lib/logging.js
159:    logger.warn('unknown log levelName "%s": cannot setLogLevel', levelName)
```

and `logger.error` cases:

```
% rg logger.error
lib/agent.js
164:    this.logger.error('Elastic APM isn\'t correctly configured: Missing serviceName')
168:    this.logger.error('Elastic APM isn\'t correctly configured: serviceName "%s" contains invalid characters! (allowed: a-z, A-Z, 0-9, _, -, <space>)', this._conf.serviceName)
172:    this.logger.error('Elastic APM isn\'t correctly configured: serverUrl "%s" contains an invalid port! (allowed: 1-65535)', this._conf.serverUrl)
257:    this.logger.error('Can\'t add filter of type %s', typeof fn)
266:    this.logger.error('Can\'t add filter of type %s', typeof fn)
275:    this.logger.error('Can\'t add filter of type %s', typeof fn)
483:    this.logger.error('Can\'t add callback of type %s', typeof callback)

lib/lambda.js
51:        if (err) agent.logger.error('Flush error:', err.message)

lib/instrumentation/transaction.js
169:    this._agent.logger.error('cannot encode un-ended transaction: %o', { trans: this.id, parent: this.parentId, trace: this.traceId })

lib/config.js
328:          agent.logger.error('APM Server transport error: %s', err.stack)
360:          agent.logger.error(msg)
650:      opts.logger.error('Elastic APM initialization error: Can\'t read server CA cert file %s (%s)', opts.serverCaCertFile, err.message)

lib/instrumentation/http-shared.js
228:            agent.logger.error('Could not set destination context:', e.message)

lib/instrumentation/index.js
101:      this._agent.logger.error('Invalid patch handler type:', type)
199:        this._agent.logger.error('Invalid patch handler type "%s" for module "%s"', type, name)
229:        agent.logger.error('error encoding span %o', { span: span.id, parent: span.parentId, trace: span.traceId, name: span.name, type: span.type, error: err.message })

lib/cloud-metadata/callback-coordination.js
37:          this.logger.error('metadata requests did not finish, possible deadlock')
77:      this.logger.error('Can not schedule callback, already started')

lib/instrumentation/modules/http2.js
192:          agent.logger.error('Could not set destination context:', e.message)

lib/instrumentation/modules/cassandra-driver.js
61:          agent.logger.error('unable to identify span exit point for cassandra-driver')
117:          agent.logger.error('unable to identify span exit point for cassandra-driver')
156:          agent.logger.error('unable to identify span exit point for cassandra-driver')
190:          agent.logger.error('unable to identify span exit point for cassandra-driver')
```

**Comment by trentm** [comment]
An open question is whether this logging format change is a negative impact on the user that we need/want/care to mitigate.  True the log *message* is by default (i.e. without `ecslog` usage) less visible, but it is still there. The ECS log output includes a "log.level" which can give the user some idea of severity that wasn't communicated before. I think that'll be helpful, for example, to change the "APM Server transport error" to be a *warning* in a subsequent change. The "log.logger=elastic-apm-node" also can help disambiguate in a noisy log that the message is from the APM agent, vs. from some part of a large app.

My inclination is this format change is fine, i.e. not something we need to mitigate for users looking at the log output on the console. However, I'm biased. @astorm do you have an opinion on this?

**Comment by astorm** [comment]
> My inclination is this format change is fine, i.e. not something we need to mitigate for users looking at the log output on the console.

That seems reasonable.  While the format has changed/become more complex, the `message` property still contains the same text that a user might be visually-scanning or `grep`-ing for.  No need for mitigation other than mentioning in the CHANGELOG that the log format has changed. 

@trentm

File: lib/agent.js [hunk 1]
```diff
@@ -459,7 +459,9 @@ Agent.prototype.flush = function (cb) {
     // TODO: Only bind the callback if the transport can't use AsyncResource from async hooks
     this._transport.flush(cb && this._instrumentation.bindFunction(cb))
   } else {
-    this.logger.warn(new Error('cannot flush agent before it is started'))
+    // Log an *err* to provide a stack for the user.
+    const err = new Error('cannot flush agent before it is started')
+    this.logger.warn({ err }, err.message)
```
Comment (astorm) [inline lib/agent.js:464]: This is interesting to me -- I didn't know about pino's optional `mergingObject` parameter/argument until [researching this syntax](https://getpino.io/#/docs/api?id=trace). (nothing to see/do here, carry on)

**Review by astorm — DISMISSED** [review]
Looks like a welcome addition to the agent.  We get ourselves out of the `console-log-level`  business, get some nice overlap with our `ecs-format-pino`, and our users get all the advantages of using one of Node's mature defacto logging libraries.  

Approving, pending CI passing after the additional work on the "placeholder %s", fixes

File: lib/agent.js (hunk 1 above)
Comment (trentm) [reply lib/agent.js:464]: It originates from https://github.com/trentm/node-bunyan#log-method-api but there are some minor differences.
- pino dropped the `log.info()`-with-no-args call signature for the more traditional log4j `log.isLevelEnabled(level)` (https://getpino.io/#/docs/api?id=islevelenabled) which was a good call.
- Bunyan treats `log.info(err)` and `log.info({err: err})` the same. Pino does not. The form that logs the error with the `err` field name is used by `@elastic/ecs-pino-format` module to format the error fields into ECS format, so we should bias to using that form.
- Bunyan will use `err.message` for the "message" field with `log.info({err: err})`. Pino will leave the message empty. Arguably pino's is cleaner, but for the (rare) case of an APM Agent user specifying a custom `logger`, they will only get the "message" field, so it is nice to have *some* message there.

File: lib/logging.js [hunk 2]
```diff
@@ -29,68 +28,142 @@
...
+    pinoLevel = 'trace'
   }
 
-  const logger = consoleLogLevel({
-    level: logApiLevel
-  })
-  logger[LOGGER_IS_OURS_SYM] = true // used for isLoggerCustom()
-  logger.level = logApiLevel // used for testing
+  if (customLogger) {
+    // Is this a pino logger? If so, it supports to API the agent requires and
```
[the remark this answers is not in the document]
Comment (trentm) [reply lib/logging.js:116]: Yes. Typo. Thanks. I'll fix that.

[the remark this answers is not in the document]
Comment (trentm) [reply lib/logging.js]: commit 174f977

**Review by astorm — DISMISSED** [review]

File: docs/configuration.asciidoc [hunk 3]
```diff
@@ -431,34 +431,66 @@ require('elastic-apm-node').start({
...
 [[logger]]
 ==== `logger`
 
 * *Type:* object
-
-Set a custom logger, e.g. https://github.com/trentm/node-bunyan[bunyan]:
+* *Env:* `ELASTIC_APM_LOGGER=false` to _ignore_ a custom logger
+
+By default, the APM agent logs to stdout in
+https://www.elastic.co/guide/en/ecs-logging/overview/current/intro.html[ecs-logging]
```
Comment (bmorelli25) [inline docs/configuration.asciidoc:450]:
```suggestion
{ecs-logging-ref}/intro.html[ecs-logging]
```

File: CHANGELOG.asciidoc [hunk 4]
```diff
@@ -36,6 +36,32 @@ Notes:
 [float]
 ===== Features
 
+* The APM agent's own internal logging now uses structured JSON logging using
+  the https://getpino.io/#/docs/api?id=logger[pino API], and formatted in
+  https://www.elastic.co/guide/en/ecs-logging/overview/current/intro.html[ecs-logging]
```
Comment (bmorelli25) [inline CHANGELOG.asciidoc:41]:
```suggestion
  {ecs-logging-ref}/intro.html[ecs-logging]
```

File: docs/troubleshooting.asciidoc [hunk 5]
```diff
@@ -114,9 +114,28 @@ Please create a new topic in the https://discuss.elastic.co/c/apm[Elastic APM di
...
+
+For example:
+
+[source,bash]
+----
+ELASTIC_APM_LOG_LEVEL=trace ELASTIC_APM_LOGGER=false node app.js | tee -a apm-debug.log
+----
+
+If you are capturing debugging output for Elastic support, for help on the
+Elastic forums, or for a GitHub issue, please *please upload the complete
```
Comment (bmorelli25) [inline docs/troubleshooting.asciidoc:136]: Not sure if the double please was intended here

```suggestion
Elastic forums, or for a GitHub issue, *please upload the complete
```

**Review by bmorelli25 — COMMENTED** [review]
This looks really good. Just a few suggestions.

File: docs/troubleshooting.asciidoc (hunk 5 above)
Comment (trentm) [reply docs/troubleshooting.asciidoc:136]: Ha ha, no it wasn't.

**Comment by trentm** [comment]
I'll assume rubberstamp re-approve by Alan and Brandon.

**Pull request merged**